/*
*  Copyright (c) 2020, The OpenThread Authors.
*  Copyright 2025 NXP
*  All rights reserved.
*
*  Redistribution and use in source and binary forms, with or without
*  modification, are permitted provided that the following conditions are met:
*  1. Redistributions of source code must retain the above copyright
*     notice, this list of conditions and the following disclaimer.
*  2. Redistributions in binary form must reproduce the above copyright
*     notice, this list of conditions and the following disclaimer in the
*     documentation and/or other materials provided with the distribution.
*  3. Neither the name of the copyright holder nor the
*     names of its contributors may be used to endorse or promote products
*     derived from this software without specific prior written permission.
*
*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
*  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
*  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
*  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
*  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
*  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
*  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
*  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
*  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
*  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
*  POSSIBILITY OF SUCH DAMAGE.
*/

/*
* Header for patch information.
* Description           : Vendor patterns strings runtime update
*                         New SetVendorPatterns API in NetworkDiagnostic and correlated otThreadSetVendorPatterns API in netdiag_api.c
*                         In Daemon::SetUp, call both otPlatRadioGetFwVersionString and otThreadSetVendorPatterns
* Version               : v001
* Mandatory/Optional    : Mandatory
* Impact if not applied : Vendor Name, Model, SW Version and App Url will remain the same as defined at buildtime
*/

Upstream-Status: Inappropriate [platform specific changes]

diff --git a/etc/cmake/options.cmake b/etc/cmake/options.cmake
index f1172095d..4197ac46a 100644
--- a/etc/cmake/options.cmake
+++ b/etc/cmake/options.cmake
@@ -335,6 +335,7 @@ ot_multi_option(OT_LOG_OUTPUT OT_LOG_OUTPUT_VALUES OPENTHREAD_CONFIG_LOG_OUTPUT
 ot_string_option(OT_VENDOR_NAME OPENTHREAD_CONFIG_NET_DIAG_VENDOR_NAME "set the vendor name config")
 ot_string_option(OT_VENDOR_MODEL OPENTHREAD_CONFIG_NET_DIAG_VENDOR_MODEL "set the vendor model config")
 ot_string_option(OT_VENDOR_SW_VERSION OPENTHREAD_CONFIG_NET_DIAG_VENDOR_SW_VERSION "set the vendor sw version config")
+ot_string_option(OT_VENDOR_APP_URL OPENTHREAD_CONFIG_NET_DIAG_VENDOR_APP_URL "set the vendor app url config")
 ot_string_option(OT_BORDER_AGENT_SERVICE_NAME OPENTHREAD_CONFIG_BORDER_AGENT_MESHCOP_SERVICE_BASE_NAME "set the border agent service base name")
 
 set(OT_POWER_SUPPLY_VALUES "BATTERY" "EXTERNAL" "EXTERNAL_STABLE" "EXTERNAL_UNSTABLE")
diff --git a/include/openthread/netdiag.h b/include/openthread/netdiag.h
index 975081b2e..04825f545 100644
--- a/include/openthread/netdiag.h
+++ b/include/openthread/netdiag.h
@@ -399,6 +399,26 @@ const char *otThreadGetVendorSwVersion(otInstance *aInstance);
  */
 const char *otThreadGetVendorAppUrl(otInstance *aInstance);
 
+/**
+ * Set all Vendor patterns strings.
+ *
+ * @p aVendorName should be UTF8 with max length of 32 chars (`MAX_VENDOR_NAME_TLV_LENGTH`). Maximum length does not
+ * @p aVendorModel should be UTF8 with max length of 32 chars (`MAX_VENDOR_MODEL_TLV_LENGTH`). Maximum length does not
+ * @p aVendorSwVersion should be UTF8 with max length of 32 chars (`MAX_VENDOR_SW_VERSION_TLV_LENGTH`). Maximum length does not
+ * @p aVendorAppUrl should be UTF8 with max length of 32 chars (`MAX_VENDOR_APP_URL_TLV_LENGTH`). Maximum length does not
+ * include the null `\0` character.
+ *
+ * @param[in] aInstance         A pointer to an OpenThread instance.
+ * @param[in] aVendorName       The vendor Name string.
+ * @param[in] aVendorModel      The vendor Model string.
+ * @param[in] aVendorSwVersion  The vendor Sw Version string.
+ * @param[in] aVendorName       The vendor AppUrl string.
+ *
+ * @retval OT_ERROR_NONE          Successfully set the vendor name.
+ * @retval OT_ERROR_INVALID_ARGS  @p aVendorName is not valid (too long or not UTF8).
+ */
+otError otThreadSetVendorPatterns(otInstance *aInstance, const char *aVendorName, const char *aVendorModel, const char *aVendorSwVersion, const char *aVendorAppUrl);
+
 /**
  * Set the vendor name string.
  *
diff --git a/src/core/api/netdiag_api.cpp b/src/core/api/netdiag_api.cpp
index 623ca53ed..53e407cff 100644
--- a/src/core/api/netdiag_api.cpp
+++ b/src/core/api/netdiag_api.cpp
@@ -91,6 +91,11 @@ const char *otThreadGetVendorAppUrl(otInstance *aInstance)
     return AsCoreType(aInstance).Get<NetworkDiagnostic::Server>().GetVendorAppUrl();
 }
 
+otError otThreadSetVendorPatterns(otInstance *aInstance, const char *aName, const char *aModel, const char *aSwVersion, const char *aAppUrl)
+{
+    return AsCoreType(aInstance).Get<NetworkDiagnostic::Server>().SetVendorPatterns(aName, aModel, aSwVersion, aAppUrl);
+}
+
 #if OPENTHREAD_CONFIG_NET_DIAG_VENDOR_INFO_SET_API_ENABLE
 otError otThreadSetVendorName(otInstance *aInstance, const char *aVendorName)
 {
diff --git a/src/core/thread/network_diagnostic.cpp b/src/core/thread/network_diagnostic.cpp
index dcd291371..e5a7e14ca 100644
--- a/src/core/thread/network_diagnostic.cpp
+++ b/src/core/thread/network_diagnostic.cpp
@@ -66,6 +66,24 @@ Server::Server(Instance &aInstance)
 #endif
 }
 
+Error Server::SetVendorPatterns(const char *aName, const char *aModel, const char *aSwVersion, const char *aAppUrl)
+{
+    static uint8_t alreadySet = 0;
+    if(!alreadySet)
+    {
+        alreadySet = 1;
+        static_assert(sizeof(aName) <= sizeof(VendorNameTlv::StringType), "VENDOR_NAME is too long");
+        static_assert(sizeof(aModel) <= sizeof(VendorModelTlv::StringType), "VENDOR_MODEL is too long");
+        static_assert(sizeof(aSwVersion) <= sizeof(VendorSwVersionTlv::StringType), "VENDOR_SW_VERSION is too long");
+        static_assert(sizeof(aAppUrl) <= sizeof(VendorAppUrlTlv::StringType), "VENDOR_APP_URL is too long");
+        if (StringCopy(mVendorName, aName, kStringCheckUtf8Encoding)) return OT_ERROR_FAILED;
+        if (StringCopy(mVendorModel, aModel, kStringCheckUtf8Encoding)) return OT_ERROR_FAILED;
+        if (StringCopy(mVendorSwVersion, aSwVersion, kStringCheckUtf8Encoding)) return OT_ERROR_FAILED;
+        if (StringCopy(mVendorAppUrl, aAppUrl, kStringCheckUtf8Encoding)) return OT_ERROR_FAILED;
+    }
+    return OT_ERROR_NONE;
+}
+
 #if OPENTHREAD_CONFIG_NET_DIAG_VENDOR_INFO_SET_API_ENABLE
 
 Error Server::SetVendorName(const char *aVendorName)
diff --git a/src/core/thread/network_diagnostic.hpp b/src/core/thread/network_diagnostic.hpp
index 887150b0f..fcb9f3eba 100644
--- a/src/core/thread/network_diagnostic.hpp
+++ b/src/core/thread/network_diagnostic.hpp
@@ -91,6 +91,22 @@ public:
      */
     explicit Server(Instance &aInstance);
 
+    /**
+     * Set all Vendor patterns strings.
+     *
+     * @param[in] aName       The vendor Name string.
+     * @param[in] aModel      The vendor Model string.
+     * @param[in] aSwVersion  The vendor Sw Version string.
+     * @param[in] aAppUrl       The vendor AppUrl string.
+     *
+     * @retval kErrorNone         Successfully set the vendor name.
+     * @retval kErrorInvalidArgs  @p aName is not valid (too long or not UTF8).
+     * @retval kErrorInvalidArgs  @p aModel is not valid (too long or not UTF8).
+     * @retval kErrorInvalidArgs  @p aSwVersion is not valid (too long or not UTF8).
+     * @retval kErrorInvalidArgs  @p aAppUrl is not valid (too long or not UTF8).
+     */
+    Error SetVendorPatterns(const char *aName, const char *aModel, const char *aSwVersion, const char *aAppUrl);
+
     /**
      * Sets the non-preferred channels value for `kNonPreferredChannels` TLV.
      *
@@ -186,10 +202,11 @@ public:
     Error SetVendorAppUrl(const char *aVendorAppUrl);
 
 #else
-    const char *GetVendorName(void) const { return kVendorName; }
-    const char *GetVendorModel(void) const { return kVendorModel; }
-    const char *GetVendorSwVersion(void) const { return kVendorSwVersion; }
-    const char *GetVendorAppUrl(void) const { return kVendorAppUrl; }
+    const char *GetVendorName(void) const { return mVendorName; }
+    const char *GetVendorModel(void) const { return mVendorModel; }
+    const char *GetVendorSwVersion(void) const { return mVendorSwVersion; }
+    const char *GetVendorAppUrl(void) const { return mVendorAppUrl; }
+
 #endif // OPENTHREAD_CONFIG_NET_DIAG_VENDOR_INFO_SET_API_ENABLE
 
 private:
@@ -264,12 +281,12 @@ private:
 
     template <Uri kUri> void HandleTmf(Coap::Message &aMessage, const Ip6::MessageInfo &aMessageInfo);
 
-#if OPENTHREAD_CONFIG_NET_DIAG_VENDOR_INFO_SET_API_ENABLE
+//#if OPENTHREAD_CONFIG_NET_DIAG_VENDOR_INFO_SET_API_ENABLE
     VendorNameTlv::StringType      mVendorName;
     VendorModelTlv::StringType     mVendorModel;
     VendorSwVersionTlv::StringType mVendorSwVersion;
     VendorAppUrlTlv::StringType    mVendorAppUrl;
-#endif
+//#endif
 
 #if OPENTHREAD_FTD
     Coap::MessageQueue mAnswerQueue;
diff --git a/src/posix/platform/daemon.cpp b/src/posix/platform/daemon.cpp
index 9cfb86660..5bc9fcd59 100644
--- a/src/posix/platform/daemon.cpp
+++ b/src/posix/platform/daemon.cpp
@@ -52,6 +52,7 @@
 #include <openthread/verhoeff_checksum.h>
 #include <openthread/platform/crypto.h>
 #include <openthread/platform/entropy.h>
+#include <openthread/netdiag.h>
 
 #if OPENTHREAD_POSIX_CONFIG_DAEMON_ENABLE
 
@@ -109,6 +110,9 @@ extern "C"
 #define OT_NXP_PLAT_MFG_FIRST_ANNEX_ID 55
 #define OT_NXP_PLAT_MFG_LAST_ANNEX_ID  108
 
+#define MAX_VENDOR_MODEL_STRING_SIZE 32     //< sizeof(VendorModelTlv::StringType)
+#define MAX_VENDOR_SWVERSION_STRING_SIZE 16 //< sizeof(VendorSwVersionTlv::StringType)
+
 static uint8_t mfgEnable = 0;
 
 // 15.4_INDEPENDENT_RESET
@@ -1120,6 +1124,41 @@ exit:
     otSysCliInitUsingDaemon(gInstance);
 #endif
 
+    char version[MAX_VERSION_STRING_SIZE] = {0};
+    char model[MAX_VENDOR_MODEL_STRING_SIZE] = {0};
+    char fwversion[MAX_VENDOR_SWVERSION_STRING_SIZE] = {0};
+    otPlatRadioGetFwVersionString(gInstance, version, MAX_VERSION_STRING_SIZE);
+    
+    const char *OtRepoSha1 = strchr(OPENTHREAD_CONFIG_NET_DIAG_VENDOR_SW_VERSION, '-');
+    if (OtRepoSha1 != NULL)
+    {
+        OtRepoSha1++; //don't want '-' here
+        size_t len = strlen(OtRepoSha1);
+        if (len >= sizeof(fwversion)) len = sizeof(fwversion) - 1;
+        strncpy(fwversion, OtRepoSha1, len);
+        fwversion[len] = '\0';
+    }
+    const char *FwRepoSha1 = strrchr(version, '-');
+    if (FwRepoSha1 != NULL)
+    {
+        size_t modelLen = FwRepoSha1 - version;
+        if (modelLen >= sizeof(model)) modelLen = sizeof(model) - 1;
+        strncpy(model, version, modelLen);
+        model[modelLen] = '\0';
+
+        size_t fwLen = strlen(FwRepoSha1);
+        size_t offset = strlen(fwversion);
+        if (offset + fwLen >= sizeof(fwversion)) fwLen = sizeof(fwversion) - offset - 1;
+        strncpy(fwversion + offset, FwRepoSha1, fwLen);
+        fwversion[offset + fwLen] = '\0';
+    }
+    
+    LogInfo("Vendor Name               : %s\r\n", OPENTHREAD_CONFIG_NET_DIAG_VENDOR_NAME);
+    LogInfo("Vendor Model              : %s\r\n", model);
+    LogInfo("Vendor Software Version   : %s\r\n", fwversion);
+    LogInfo("Vendor App Url            : %s\r\n", OPENTHREAD_CONFIG_NET_DIAG_VENDOR_APP_URL);
+    otThreadSetVendorPatterns(gInstance, OPENTHREAD_CONFIG_NET_DIAG_VENDOR_NAME, model, fwversion, OPENTHREAD_CONFIG_NET_DIAG_VENDOR_APP_URL);
+
     otCliSetUserCommands(kCommands, OT_ARRAY_LENGTH(kCommands), gInstance);
 
     Mainloop::Manager::Get().Add(*this);
